system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.hpp

```diff
@@ -160,10 +219,165 @@ class SubscriberTopicStatisticsNode : public system_metrics_collector::MetricsMe
+    collect_topic_desc);
+  validateStringParam(collect_topic_name_);
+
+  const auto publish_topic_desc = buildTopicParameterDescriptor(
+    "The topic to publish topic statistics to.");
+  publish_topic_name_ = declare_parameter(
+    topic_statistics_constants::kPublishStatsTopicNameParam,
+    system_metrics_collector::collector_node_constants::kStatisticsTopicName,
+    publish_topic_desc);
+  validateStringParam(publish_topic_name_);
+
+  initializeCollectors(statistics_collectors_);
+
+  auto callback = [this](typename T::SharedPtr received_message) {
+      for (const auto & collector : statistics_collectors_) {
+        collector->OnMessageReceived(*received_message, this->LifecycleNode::now().nanoseconds());
+      }
+    };
+  subscription_ = create_subscription<T>(collect_topic_name_, 10, callback);
+}
+
+template<typename T>
+void SubscriberTopicStatisticsNode<T>::StartPublisher()
+{
+  rcpputils::check_true(publish_timer_ == nullptr);
+
+  if (publisher_ == nullptr) {
+    publisher_ = create_publisher<MetricsMessage>(publish_topic_name_, 10);
+  }
+
+  publisher_->on_activate();
+
+  RCLCPP_DEBUG(this->get_logger(), "SetupStart: creating publish_timer_");
+  publish_timer_ = this->create_wall_timer(
+    publish_period_, [this]() {
+      this->PublishStatisticMessage();
+      this->ClearCollectorMeasurements();
+    });
+  window_start_ = now();
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_activate(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_activate");
+  for (const auto & collector : statistics_collectors_) {
+    if (!collector->Start()) {
+      return CallbackReturn::ERROR;
+    }
+  }
+
+  StartPublisher();
+  return CallbackReturn::SUCCESS;
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_deactivate(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_deactivate");
+  for (const auto & collector : statistics_collectors_) {
+    if (!collector->Stop()) {
+      return CallbackReturn::ERROR;
+    }
+  }
+
+  StopPublisher();
+  return CallbackReturn::SUCCESS;
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_shutdown(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_shutdown");
+  ClearCollectorMeasurements();
+  publisher_.reset();
+
+  return CallbackReturn::SUCCESS;
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_error(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_error");
+  ClearCollectorMeasurements();
+  if (publisher_ != nullptr) {
+    publisher_.reset();
+  }
+
+  return CallbackReturn::SUCCESS;
+}
+
+template<typename T>
```

Comment:
Is ```template<typename T>``` really needed for all of the template class methods? They shouldn't need to be template functions.

Reply:
They're members functions of the class template, that's why it's needed.

Reply:
If you define inline (comment above) then extra template verbiage [isn't needed](https://blog.feabhas.com/2014/06/template-classes/).